<a href="https://colab.research.google.com/github/AdityaPareek7412/Stock_Price_Prediction_By_LLMs/blob/main/M1_Assignment_Stock_Price_Prediction_with_LLM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Module 1 No Code/Low Code Assignment: Stock Price Prediction with LLM (Using Gemini in Colab)

This assignment explores how large language models (LLMs) can perform complex tasks that previously required coding. Instead of manually writing code, we will interact with Google's Gemini (Click on the Gemini icon on the top right corner) in Google Colab to:

- Load and analyze stock price data.
- Engineer relevant features for prediction.
- Build and evaluate models such as regression and random forest.
- Discuss potential trading strategies based on predictions.

**Note**: While this notebook contains example code, students are encouraged to use LLMs to generate code snippets and explanations, managing interactions within the LLM's context window.

## Task Breakdown

### Step 1: Data Exploration
- **Prompt the LLM**: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance.
- **Key Considerations**: Ensure the model retains context while summarizing key statistics.

### Step 2: Feature Engineering
- **Prompt the LLM**: Generate meaningful features such as moving averages, technical indicators, and volatility measures.
- **Managing Context**: If the LLM forgets prior outputs, remind it by reintroducing key statistics.

### Step 3: Model Building
- **Prompt the LLM**: Train a model (regression for price prediction, random forest for trend classification).
- **Split Ratio Consideration**: Explain the importance of train-test split (e.g., 70/30) and verify the LLM respects it.

### Step 4: Model Evaluation & Prediction
- **Prompt the LLM**: Use the trained model to predict stock prices for 2023-2024.
- **Analyze Performance**: Interpret accuracy, R² scores, and classification reports.

### Step 5: Strategy Discussion
- **Prompt the LLM**: Based on model performance, discuss possible trading strategies.
- **Critical Thinking**: Evaluate whether the model is truly useful for financial decision-making.

---

Throughout the assignment, keep in mind that LLMs have a limited context window. If necessary, break down tasks into smaller chunks and reintroduce critical information.



## Using Gemini in Colab

To use Gemini directly in your browser within Google Colab:

1. **Create a new code cell** by clicking the `+ Code` button.
2. In the newly created cell, you will see a placeholder text: **"Start coding or generate with AI"**.
3. Click on **"Generate"** and enter your prompt, such as:
   > "Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics."
4. Gemini will generate code based on your prompt. You can then review, modify, and run it within the notebook.
5. If the model loses context, you may need to remind it of previous responses or break down tasks into smaller steps.

This method allows you to interactively generate and refine code without manually writing every line.


In [5]:
# prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics.

!pip install yfinance

import yfinance as yf
import pandas as pd

# Download historical data for AAPL
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Display summary statistics
print(data.describe())


/tmp/ipython-input-1433581267.py:9: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.488414  129.035076  125.828629  127.410552  1.120920e+08
std      30.386976   30.664098   30.105383   30.415044  5.602586e+07
min      54.316933   55.316751   51.469998   55.215075  3.519590e+07
25%     112.581739  114.075470  111.125703  112.761683  7.636470e+07
50%     132.859985  133.991766  130.902078  132.681703  9.493580e+07
75%     148.053478  149.219241  146.111085  147.421458  1.296327e+08
max     178.443161  179.354917  175.609770  179.050994  4.265100e+08


In [6]:
# prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics.

!pip install yfinance

import yfinance as yf
import pandas as pd

# Download historical data for AAPL from 2020-01-01 to 2023-01-01
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Summarize key statistics
print(data.describe())


/tmp/ipython-input-730687086.py:9: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.488414  129.035076  125.828629  127.410552  1.120920e+08
std      30.386976   30.664098   30.105383   30.415044  5.602586e+07
min      54.316933   55.316751   51.469998   55.215075  3.519590e+07
25%     112.581739  114.075470  111.125703  112.761683  7.636470e+07
50%     132.859985  133.991766  130.902078  132.681703  9.493580e+07
75%     148.053478  149.219241  146.111085  147.421458  1.296327e+08
max     178.443161  179.354917  175.609770  179.050994  4.265100e+08


In [7]:
# Prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01
# using Yahoo Finance and summarize its key statistics.

import yfinance as yf

data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Summarize key statistics
print(data.describe())



/tmp/ipython-input-3699561310.py:6: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.488414  129.035076  125.828629  127.410552  1.120920e+08
std      30.386976   30.664098   30.105383   30.415044  5.602586e+07
min      54.316933   55.316751   51.469998   55.215075  3.519590e+07
25%     112.581739  114.075470  111.125703  112.761683  7.636470e+07
50%     132.859985  133.991766  130.902078  132.681703  9.493580e+07
75%     148.053478  149.219241  146.111085  147.421458  1.296327e+08
max     178.443161  179.354917  175.609770  179.050994  4.265100e+08


## Now try doing step 2 - step 5 by yourself

The sample prompts and LLM responses are listed below for your reference.

# Sample Prompts and LLM Responses
---

## Step 1: Data Exploration

### Prompt:
> Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 from Yahoo Finance, keep it in memory and summarize its key statistics.

### LLM Response:
- **Highest Closing Price:** \$182.01 on January 3, 2022
- **Lowest Closing Price:** \$56.01 on March 23, 2020
- **Average Closing Price:** \$126.69
- **Total Trading Days:** 756

---

## Step 2: Feature Engineering

### Prompt:
> Suggest some features that might help improve a stock price prediction model.

### LLM Response:
1. **Historical Prices**: Include past closing, opening, high, and low prices.
2. **Volume**: Daily trading volumes.
3. **Moving Averages**: Simple and exponential moving averages (e.g., 5-day, 10-day, 50-day).
4. **Technical Indicators**: RSI, MACD, Bollinger Bands, etc.
5. **Volatility**: Historical volatility, Average True Range (ATR).
6. **Sentiment Analysis**: News articles, social media sentiment.
7. **Economic Indicators**: Interest rates, inflation rates, GDP growth.
8. **Time-based Features**: Day of the week, month, year.
9. **Sector Performance**: Comparison with industry and sector indices.
10. **External Events**: Major geopolitical events, earnings reports.

---

## Step 3.1: Model Building - Linear Regression

### Prompt:
> Adopt the first four kinds of features and train a regression model to predict tomorrow's stock price. Split your training and testing data by a 70/30 ratio and report your performance metric.

### LLM Response:
- **Mean Squared Error (MSE):** 859.04
- **R-squared (R²) Score:** -0.009

These metrics indicate that the model is not performing well, as an R² score close to zero or negative suggests that the model does not explain the variability in the target variable better than a simple mean of the target variable. This outcome suggests that the chosen features and model might need adjustments or improvements to enhance prediction accuracy. ​​

---
(It might be too hard to predict the exact price but how about ups and downs?)
# Step 3.2: Model Building - Random Forest Classification

### Prompt:
> Create a binary target variable based on historical prices (1 if the close price increase tomorrow, 0 other), and train a Random Forest model using the same train/test split ratio. Report feature importance and performance metrics.

### LLM Response:
- **Accuracy:** 71.82%
- **Confusion Matrix:** [[84, 26],
 [36, 74]]
- **Classification Report:**

              precision    recall  f1-score   support

         0       0.70      0.76      0.73       110
         1       0.74      0.67      0.70       110
        accuracy                     0.72       220
        macro avg    0.72      0.72      0.72       220
        weighted avg  0.72      0.72      0.72       220



- **Feature Importance:** The most important feature for predicting whether the price will increase or decrease the next day is the closing price, followed by the 5-day moving average (MA_5). The model achieved an accuracy of approximately 71.82%, indicating a reasonable performance for this classification task.
         Feature	Importance
         Open	   0.0900
         High	   0.0883
         Low	   0.0980
         Close	   0.3087
         Volume	    0.0982
         MA_5	   0.1242
         MA_10	   0.0974
         MA_50	   0.0951

---

## Step 4: Model Evaluation & Prediction
### Prompt:
> Use this model to predict stock price increase/decrease for period starting from 2023-01-01 to 2024-01-01. Report your performance metrics.

### LLM Response:
The Random Forest classifier's performance on the new test data from 2023-01-01 to 2024-01-01 is as follows:
- **Accuracy:** 60.85%
- **Confusion Matrix:** [[105,   0],
 [ 83,  24]]
- **Classification Report:**

            precision    recall  f1-score   support

       0       0.56      1.00      0.72       105
       1       1.00      0.22      0.37       107

        accuracy                   0.61       212
        macro avg  0.78     0.61       0.54       212
        weighted avg 0.78    0.61        0.54       212

---
## Step 5: Strategy Discussion
### Prompt:
> What do you think about the performance of the model and what trading strategies could I build with this model?

### LLM Response:
- The model shows a moderate performance with an accuracy of approximately 60.85%. However, the precision and recall metrics indicate that while the model is good at identifying days when the stock price does not increase (class 0), it struggles significantly to identify days when the stock price does increase (class 1). This suggests a need for further feature engineering or model tuning to improve the prediction performance for this specific classification task.
- Potential Trading Strategies
Given the model's performance, potential trading strategies could be formulated as follows:

#### Conservative Strategy:

- **Focus**: Only trade on days when the model predicts a price increase with high confidence.
- **Action**: Buy stocks when the model predicts an increase (class 1).
- **Risk Management**: Since the model has a high precision for the no-increase class (class 0), it might be more reliable to stay out of the market when the model predicts no increase.

#### Contrarian Strategy:

- **Focus**: Take advantage of the model's high recall for no price increase predictions (class 0).
- **Action**: Sell short or avoid buying stocks on days when the model predicts no increase.
- **Risk Management**: This strategy assumes that the model’s predictions of no price increase are more reliable. It requires a careful assessment of market conditions to avoid false signals.

#### Hybrid Strategy:

- **Combine**: Use both conservative and contrarian strategies based on confidence thresholds.
- **Action**: Set a higher threshold for buying (only buy when the predicted probability of price increase is significantly high) and a lower threshold for selling or staying out of the market.
- **Risk Management**: Adjust thresholds dynamically based on recent model performance and market volatility.


---
## Conclusion

Through this assignment, we have explored how LLMs can assist in complex machine learning tasks without requiring extensive coding knowledge. Key takeaways:

- **Context Management**: LLMs have limited memory, so breaking tasks into smaller steps and reinforcing context is crucial.
- **Feature Engineering with LLMs**: The model can suggest features but needs guidance on financial relevance.
- **Modeling with LLMs**: LLMs can write and refine code, but human judgment is necessary to validate results.
- **Real-world Limitations**: Stock price prediction remains challenging; models should be used critically and not solely relied upon for trading decisions.

Next Steps:
- Experiment with different LLMs and compare outputs.
- Try refining feature engineering techniques.
- Test alternative modeling approaches, such as deep learning or ensemble methods.



In [9]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
import pandas as pd
import yfinance as yf

# Download historical data for AAPL from 2020-01-01 to 2023-01-01
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Feature Engineering
# Calculate Moving Averages
data['MA_5'] = data['Close'].rolling(window=5).mean()
data['MA_10'] = data['Close'].rolling(window=10).mean()
data['MA_50'] = data['Close'].rolling(window=50).mean()

# Calculate Volatility (Rolling Standard Deviation)
data['Volatility'] = data['Close'].rolling(window=14).std()

# Create the target variable (1 if price increases the next day, 0 otherwise)
data['Price_Increase'] = (data['Close'].shift(-1) > data['Close']).astype(int)

# Drop rows with NaN values created by rolling windows and the target variable
data.dropna(inplace=True)

# Define features (X) and target (y)
# Using the first four kinds of features as per the prompt's suggestion in 3.1,
# plus the moving averages and volatility already engineered
features = ['Open', 'High', 'Low', 'Close', 'Volume', 'MA_5', 'MA_10', 'MA_50', 'Volatility']
X = data[features]
y = data['Price_Increase']

# Split data into training and testing sets (70/30 ratio)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Initialize and train the Random Forest Classifier
model = RandomForestClassifier(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = model.predict(X_test)

# Report performance metrics
print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))

# Report feature importance
feature_importances = pd.Series(model.feature_importances_, index=features).sort_values(ascending=False)
print("\nFeature Importance:\n", feature_importances)

/tmp/ipython-input-3006220704.py:8: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed


Accuracy: 0.4788732394366197

Confusion Matrix:
 [[38 66]
 [45 64]]

Classification Report:
               precision    recall  f1-score   support

           0       0.46      0.37      0.41       104
           1       0.49      0.59      0.54       109

    accuracy                           0.48       213
   macro avg       0.48      0.48      0.47       213
weighted avg       0.48      0.48      0.47       213


Feature Importance:
 Volatility    0.150326
MA_50         0.118539
High          0.114947
MA_10         0.111610
Volume        0.105662
Close         0.104593
MA_5          0.100667
Open          0.099463
Low           0.094194
dtype: float64


In [10]:
# Step 4: Model Evaluation & Prediction

# Download historical data for AAPL for the prediction period
prediction_data = yf.download("AAPL", start="2023-01-01", end="2024-01-01")

# Apply the same feature engineering steps to the new data
prediction_data['MA_5'] = prediction_data['Close'].rolling(window=5).mean()
prediction_data['MA_10'] = prediction_data['Close'].rolling(window=10).mean()
prediction_data['MA_50'] = prediction_data['Close'].rolling(window=50).mean()
prediction_data['Volatility'] = prediction_data['Close'].rolling(window=14).std()

# Drop rows with NaN values created by rolling windows
prediction_data.dropna(inplace=True)

# Create the target variable for the prediction period (for evaluation)
prediction_data['Price_Increase'] = (prediction_data['Close'].shift(-1) > prediction_data['Close']).astype(int)

# Drop the last row as it will have a NaN for Price_Increase
prediction_data.dropna(inplace=True)


# Define features (X_predict) and target (y_predict) for the prediction period
X_predict = prediction_data[features]
y_predict = prediction_data['Price_Increase']

# Make predictions on the new data
y_pred_predict = model.predict(X_predict)

# Report performance metrics on the new data
print("Performance on prediction period (2023-01-01 to 2024-01-01):")
print("Accuracy:", accuracy_score(y_predict, y_pred_predict))
print("\nConfusion Matrix:\n", confusion_matrix(y_predict, y_pred_predict))
print("\nClassification Report:\n", classification_report(y_predict, y_pred_predict))

/tmp/ipython-input-1670818487.py:4: FutureWarning: YF.download() has changed argument auto_adjust default to True
  prediction_data = yf.download("AAPL", start="2023-01-01", end="2024-01-01")
[*********************100%***********************]  1 of 1 completed


Performance on prediction period (2023-01-01 to 2024-01-01):
Accuracy: 0.472636815920398

Confusion Matrix:
 [[ 90   2]
 [104   5]]

Classification Report:
               precision    recall  f1-score   support

           0       0.46      0.98      0.63        92
           1       0.71      0.05      0.09       109

    accuracy                           0.47       201
   macro avg       0.59      0.51      0.36       201
weighted avg       0.60      0.47      0.33       201



In [11]:
import yfinance as yf
import pandas as pd

# Download historical data for AAPL from 2020-01-01 to 2023-01-01
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Feature Engineering
# Calculate Moving Averages
data['MA_5'] = data['Close'].rolling(window=5).mean()
data['MA_10'] = data['Close'].rolling(window=10).mean()
data['MA_50'] = data['Close'].rolling(window=50).mean()

# Calculate Volatility (Rolling Standard Deviation)
data['Volatility'] = data['Close'].rolling(window=14).std()

# Drop rows with NaN values created by rolling windows
data.dropna(inplace=True)
# Display the first few rows with new features
print(data.head())

/tmp/ipython-input-2431956694.py:5: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed

Price           Close       High        Low       Open     Volume       MA_5  \
Ticker           AAPL       AAPL       AAPL       AAPL       AAPL              
Date                                                                           
2020-03-13  67.292786  67.764857  61.235780  64.126297  370732000  65.515375   
2020-03-16  58.635769  62.719763  58.100756  58.572824  322423600  64.355299   
2020-03-17  61.213978  62.363885  57.713406  59.918814  324056000  62.782706   
2020-03-18  59.715469  60.521618  57.403543  58.045074  300233600  61.390224   
2020-03-19  59.257923  61.209139  58.732596  59.889769  271857200  61.223185   

Price           MA_10      MA_50 Volatility  
Ticker                                       
Date                                         
2020-03-13  68.412665  73.941989   3.483190  
2020-03-16  67.042458  73.663934   4.311666  
2020-03-17  66.159809  73.451548   4.512722  
2020-03-18  64.802431  73.197744   4.904133  
2020-03-19  63.637027  72.941599   5.

In [12]:
import yfinance as yf
import pandas as pd

# Download historical data for AAPL from 2020-01-01 to 2023-01-01
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Feature Engineering
# Calculate Moving Averages
data['MA_5'] = data['Close'].rolling(window=5).mean()
data['MA_10'] = data['Close'].rolling(window=10).mean()
data['MA_50'] = data['Close'].rolling(window=50).mean()

# Calculate Volatility (Rolling Standard Deviation)
data['Volatility'] = data['Close'].rolling(window=14).std()

# Drop rows with NaN values created by rolling windows
data.dropna(inplace=True)

# Display the first few rows with new features
print(data.head())

/tmp/ipython-input-790055927.py:5: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed

Price           Close       High        Low       Open     Volume       MA_5  \
Ticker           AAPL       AAPL       AAPL       AAPL       AAPL              
Date                                                                           
2020-03-13  67.292786  67.764857  61.235780  64.126297  370732000  65.515375   
2020-03-16  58.635769  62.719763  58.100756  58.572824  322423600  64.355299   
2020-03-17  61.213978  62.363885  57.713406  59.918814  324056000  62.782706   
2020-03-18  59.715469  60.521618  57.403543  58.045074  300233600  61.390224   
2020-03-19  59.257923  61.209139  58.732596  59.889769  271857200  61.223185   

Price           MA_10      MA_50 Volatility  
Ticker                                       
Date                                         
2020-03-13  68.412665  73.941989   3.483190  
2020-03-16  67.042458  73.663934   4.311666  
2020-03-17  66.159809  73.451548   4.512722  
2020-03-18  64.802431  73.197744   4.904133  
2020-03-19  63.637027  72.941599   5.